osmoutils/partialord/internal/dag/dag.go

```diff
@@ -0,0 +1,305 @@
+	}
+	if len(nodeNames) != len(nodes) {
+		panic("provided multiple nodes with the same name")
+	}
+	return Dag{
+		directedEdgeList: directedEdgeList,
+		nodeNames:        nodeNames,
+		idToNodeNames:    idToNodeNames,
+	}
+}
+
+// Copy returns a new dag struct that is a copy of the original dag.
+// Edges can be mutated in the copy, without altering the original.
+func (dag Dag) Copy() Dag {
+	directedEdgeList := make([]map[int]int8, len(dag.nodeNames))
+	for i := 0; i < len(dag.nodeNames); i++ {
+		originalEdgeList := dag.directedEdgeList[i]
+		directedEdgeList[i] = make(map[int]int8, len(originalEdgeList))
+		for k, v := range originalEdgeList {
+			directedEdgeList[i][k] = v
+		}
+	}
+	// we re-use nodeNames and idToNodeNames as these are fixed at dag creation.
+	return Dag{
+		directedEdgeList: directedEdgeList,
+		nodeNames:        dag.nodeNames,
+		idToNodeNames:    dag.idToNodeNames,
+	}
+}
+
+func (dag Dag) hasDirectedEdge(u, v int) bool {
+	uAdjacencyList := dag.directedEdgeList[u]
+	_, exists := uAdjacencyList[v]
+	return exists
+}
+
+// addEdge adds a directed edge from u -> v.
+func (dag *Dag) addEdge(u, v int) error {
+	if u == v {
+		return fmt.Errorf("can't make self-edge")
+	}
+	if dag.hasDirectedEdge(v, u) {
+		return fmt.Errorf("dag has conflicting edge")
+	}
+	dag.directedEdgeList[u][v] = 1
+	dag.directedEdgeList[v][u] = -1
+	return nil
+}
+
+// replaceEdge adds a directed edge from u -> v.
+// it removes any edge that may already exist between the two.
+func (dag *Dag) replaceEdge(u, v int) error {
+	if u == v {
+		return fmt.Errorf("can't make self-edge")
+	}
+
+	dag.directedEdgeList[u][v] = 1
+	dag.directedEdgeList[v][u] = -1
+	return nil
+}
+
+// deleteEdge deletes edges between u and v.
+func (dag *Dag) deleteEdge(u, v int) {
+	delete(dag.directedEdgeList[u], v)
+	delete(dag.directedEdgeList[v], u)
+}
+
+func (dag *Dag) AddEdge(u, v string) error {
+	uIndex, uExists := dag.nodeNames[u]
+	vIndex, vExists := dag.nodeNames[v]
+	if !uExists || !vExists {
+		return fmt.Errorf("one of %s, %s does not exist in dag", u, v)
+	}
+	return dag.addEdge(uIndex, vIndex)
+}
+
+// ReplaceEdge adds a directed edge from u -> v.
+// it removes any edge that may already exist between the two.
+func (dag *Dag) ReplaceEdge(u, v string) error {
+	uIndex, uExists := dag.nodeNames[u]
+	vIndex, vExists := dag.nodeNames[v]
+	if !uExists || !vExists {
+		return fmt.Errorf("one of %s, %s does not exist in dag", u, v)
+	}
+	return dag.replaceEdge(uIndex, vIndex)
+}
+
+// AddFirstElements sets the provided elements to be first in all orderings.
+// So if were making an ordering over {A, B, C, D, E}, and elems provided is {D, B, A}
+// then we are guaranteed that the total ordering will begin with {D, B, A}
+func (dag *Dag) AddFirstElements(nodes ...string) error {
+	nodeIds, err := dag.namesToIds(nodes)
+	if err != nil {
+		return err
+	}
+
+	return dag.addFirst(nodeIds)
+}
+
+func (dag *Dag) addFirst(nodes []int) error {
```

Comment:
Not sure this helps readability of the function logic. https://dave.cheney.net/2019/01/29/you-shouldnt-name-your-variables-after-their-types-for-the-same-reason-you-wouldnt-name-your-pets-dog-or-cat

Reply:
I think your reply was in respone to [another](https://github.com/osmosis-labs/osmosis/pull/1335/files/b53df15ffcbf38e781df4f13256b1e8e73a8f063#r857060889) comment?

I suggested the name `nodeIds` here because nodes may have a:
1. name
2. id

When I was reading this, I got stumbled because `nodes` has two meanings. Although it's implied that the type of `nodes` refers to option 2, it got me thinking whether `nodes` may I have a third integer meaning

Reply:
I tried the change before posting that, and found the code / comments concretely harder to parse personally